### PR TITLE
Pt 157273931 channel reestablish

### PR DESCRIPTION
**Pull request by uwiger**
See https://www.pivotaltracker.com/story/show/157273931

* leave a channel (mutual handshake, no signing), withoug chain activity
* establish a new channel instance based on latest mutually signed offchain state
* sliding event window in the channel fsm for debugging purposes
* Erlang API for leave/reestablish/change_config/get_history
* Only config change supported, at first, is `log_keep`, which controls the size of the sliding window
* Test cases for leave+reestablish, leave+reestablish+update+close_mutual, and for config change + get_history

### CODE REVIEW

**Comment by uwiger** [comment]
Oh, and a off-chain state cache was added. This is used to recover the full state trees at reestablish.
One thing that remains is that the cache needs to be garbage-collected.

File: apps/aechannel/src/aesc_codec.hrl [hunk 1]
```diff
@@ -18,27 +19,32 @@
...
--define(ID_UPDATE_ERR   , 9).
--define(ID_DEP_CREATED  , 10).
--define(ID_DEP_SIGNED   , 11).
--define(ID_DEP_LOCKED   , 12).
--define(ID_DEP_ERR      , 13).
--define(ID_WDRAW_CREATED, 14).
--define(ID_WDRAW_SIGNED , 15).
--define(ID_WDRAW_LOCKED , 16).
--define(ID_WDRAW_ERR    , 17).
+-define(ID_CH_REEST_ACK , 4).
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_codec.hrl]: Some ids need to slide at https://github.com/aeternity/protocol/blame/52de7f9a1e86428a9d6b729d72764f088372c440/channels/OFF-CHAIN.md#L71 too?

File: apps/aechannel/src/aesc_codec.hrl [hunk 2]
```diff
@@ -18,27 +19,32 @@
...
+-define(ID_UPDATE_ERR   , 10).
+-define(ID_DEP_CREATED  , 11).
+-define(ID_DEP_SIGNED   , 12).
+-define(ID_DEP_LOCKED   , 13).
+-define(ID_DEP_ERR      , 14).
+-define(ID_WDRAW_CREATED, 15).
+-define(ID_WDRAW_SIGNED , 16).
+-define(ID_WDRAW_LOCKED , 17).
+-define(ID_WDRAW_ERR    , 18).
+-define(ID_LEAVE        , 94).
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_codec.hrl]: Does this require spec at https://github.com/aeternity/protocol/blame/52de7f9a1e86428a9d6b729d72764f088372c440/channels/OFF-CHAIN.md ?

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 3]
```diff
@@ -22,6 +22,9 @@
 -export([ new/1                       %%  (Opts) -> {ok, Tx, State}
         , check_initial_update_tx/3   %%  (SignedTx, State, Opts)
         , check_update_tx/3           %%  (SignedTx, State, Opts)
+        , check_reestablish_tx/2      %%  (SignedTx, State) -> {ok,NewSt} | error()
+        , is_latest_signed_tx/2       %%  (SignedTx, State) -> bookean()
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_offchain_state.erl]: boolean

File: apps/aechannel/src/aesc_state_cache.erl [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        _ ->
+            ok
+    end,
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    {ok, #st{}}.
+
+handle_call({new, ChId, PubKey, State}, _From, St) ->
+    case ets:insert_new(?TAB, #ch{id = key(ChId, PubKey), state = State}) of
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_state_cache.erl]: I see memory based. Maybe worth a metric so operator has feeling of memory used by this?

**Comment by lucafavatella** [comment]
I am taking the liberty to label as WIP as maybe @uwiger wanted to include some amendments.

**Comment by lucafavatella** [comment]
@uwiger Can this be closed in favour of https://github.com/aeternity/epoch/pull/1378 ?

**Comment by uwiger** [comment]
Superseded by #1378

**Pull request closed without merging**